File: cs/src/ConstraintSolver/Intervals/UpwardPropagator.cs
//#define DEBUG_UP
using System;
using System.IO;
using System.Threading;
using System.Collections.Generic;
//using Alica.Reasoner;
//using Al=Alica;
using AutoDiff;


namespace Alica.Reasoner.IntervalPropagation
{
		
	public class UpwardPropagator : ITermVisitor<bool>
	{
		internal TermList Changed;
		//internal Queue<Term> Changed;
		/*private void AddAllChanged(List<Term> l) {
			foreach(Term t in l) {
				if(!Changed.Contains(t)) Changed.Enqueue(t);
			}
		}*/
		private void AddChanged(Term t) {			
			/*foreach(Term s in t.Parents) {
				Changed.Enqueue(s);
			}
			Changed.Enqueue(t);*/
			foreach(Term s in t.Parents) {
				//Changed.Enqueue(s);
				if(!Changed.Contains(s)) Changed.Enqueue(s);
				//Changed.MoveToEnd(s);
			}
			if(!Changed.Contains(t)) Changed.Enqueue(t);
			//Changed.Enqueue(t);
		}
		public UpwardPropagator() {
		}
		
		internal DownwardPropagator DP {get; set;}
		
		
        public bool Visit(Constant constant) {
			
			return false;
		}

        public bool Visit(Zero zero) {
			
			return false;
		}

        public bool Visit(ConstPower intPower) {
			bool includesZero = intPower.Base.Min*intPower.Base.Max <= 0;	
			if (intPower.Exponent > 0) {
				double a = Math.Pow(intPower.Base.Max,intPower.Exponent);
				double b = Math.Pow(intPower.Base.Min,intPower.Exponent);
				if (includesZero) {
					if(UpdateInterval(intPower,Math.Min(0,Math.Min(a,b)),Math.Max(0,Math.Max(a,b)))) {
					//if(UpdateInterval(intPower,Math.Min(0,Math.Pow(intPower.Base.Min,intPower.Exponent)),Math.Max(0,Math.Pow(intPower.Base.Max,intPower.Exponent)))) {
						AddChanged(intPower);
						return true;
					}
				} else {
					if(UpdateInterval(intPower,Math.Min(a,b),Math.Max(a,b))) {
					//if(UpdateInterval(intPower,Math.Pow(intPower.Base.Min,intPower.Exponent),Math.Pow(intPower.Base.Max,intPower.Exponent))) {
						AddChanged(intPower);
						return true;
					}
				}				
			} 
			else if (!includesZero) {
				double a = Math.Pow(intPower.Base.Max,intPower.Exponent);
				double b = Math.Pow(intPower.Base.Min,intPower.Exponent);

				//Console.WriteLine("Cur: {0} [{1} : {2}]",intPower,intPower.Min,intPower.Max);
				//Console.WriteLine("Base: [{0} : {1}]",intPower.Base.Min,intPower.Base.Max);				
				
				if(UpdateInterval(intPower,Math.Min(a,b),Math.Max(a,b))) {
					//Console.WriteLine("From UW intpower {0}",intPower.Exponent);
				//if(UpdateInterval(intPower,Math.Pow(intPower.Base.Max,intPower.Exponent),Math.Pow(intPower.Base.Min,intPower.Exponent))) {
					AddChanged(intPower);
					return true;
				}
			} //else +- Infinity is possible
			return false;
		}
		public bool Visit(TermPower tp) {
			throw new NotImplementedException("Propagation for TemPower not implemented");					
		}
		public bool Visit(Gp gp) {
			throw new NotImplementedException("Propagation for TemPower not implemented");					
		}

        public bool Visit(Product product) {
			double aa = product.Left.Min*product.Right.Min;
			double bb = product.Left.Max*product.Right.Max;
			double max;
			double min;
			if (product.Left == product.Right) {
				min = Math.Min(aa,bb);
				max = Math.Max(aa,bb);
				if (product.Left.Min*product.Left.Max <= 0) min = 0;								
			}
			else {
				double ab = product.Left.Min*product.Right.Max;
				double ba = product.Left.Max*product.Right.Min;
				max = Math.Max(aa,Math.Max(ab,Math.Max(ba,bb)));
				min = Math.Min(aa,Math.Min(ab,Math.Min(ba,bb)));				
			}			
			if(UpdateInterval(product,min,max)) {	
				AddChanged(product);
				return true;
			}			
			return false;			
		}

        public bool Visit(Sigmoid sigmoid) {			
			throw new NotImplementedException("Sigmoidal propagation not implemented");
		}
		
		public bool Visit(LinSigmoid sigmoid) {			
			throw new NotImplementedException("Sigmoidal propagation not implemented");
		}
		
        public bool Visit(LTConstraint constraint) {
			if (constraint.Left.Max < constraint.Right.Min) {
				if(UpdateInterval(constraint,1,1)) {
					AddChanged(constraint);
					return true;
				}
			}
			else if (constraint.Left.Min >= constraint.Right.Max) {
				//Console.WriteLine("LT UP negated: {0} {1}",constraint.Left.Min ,constraint.Right.Max);
				if(UpdateInterval(constraint,Double.NegativeInfinity,0)) {
					AddChanged(constraint);
					return true;
				}
			}
			return false;
		}
		
        public bool Visit(LTEConstraint constraint) {			
			if (constraint.Left.Max <= constraint.Right.Min) {
				if(UpdateInterval(constraint,1,1)) {
					AddChanged(constraint);
					return true;
				}
			}
			else if (constraint.Left.Min > constraint.Right.Max) {
				if(UpdateInterval(constraint,Double.NegativeInfinity,0)) {
					AddChanged(constraint);
					return true;
				}
			}

			return false;
		}
		
        public bool Visit(Min min) {
			if(UpdateInterval(min,Math.Min(min.Left.Min,min.Right.Min),Math.Max(min.Left.Max,min.Right.Max))) {
				AddChanged(min);
				return true;
			}
			return false;
		}
		
        public bool Visit(Max max) {
			if(UpdateInterval(max,Math.Min(max.Left.Min,max.Right.Min),Math.Max(max.Left.Max,max.Right.Max))) {
				AddChanged(max);
				return true;
			}
			return false;
		}
		
        public bool Visit(And and) {			
			if(and.Left.Min > 0 && and.Right.Min > 0) {
				if(UpdateInterval(and,1,1)) {
					AddChanged(and);
					return true;
				}
			}	
			else if(and.Left.Max <= 0 || and.Right.Max <= 0) {
				if(UpdateInterval(and,Double.NegativeInfinity,0)) {
					AddChanged(and);
					return true;
				}
			}
			return false;
		}
		
        public bool Visit(Or or) {
			if(or.Left.Min > 0 || or.Right.Min > 0) {
				if(UpdateInterval(or,1,1)) {
					AddChanged(or);
					return true;
				}
			}	
			else if(or.Left.Max <= 0 && or.Right.Max <= 0) {
				if(UpdateInterval(or,Double.NegativeInfinity,0)) {
					AddChanged(or);
					return true;
				}
			}			
			return false;
		}
		
        public bool Visit(ConstraintUtility cu) {
			if (cu.Constraint.Max < 1) {
				if (UpdateInterval(cu,Double.NegativeInfinity,cu.Constraint.Max)) {
					AddChanged(cu);
					return true;
				}
			}
			if (UpdateInterval(cu,Double.NegativeInfinity,cu.Utility.Max)) {
				AddChanged(cu);
				return true;
			}
			return false;
		}
		
		public bool Visit(Reification reif) {
			if (reif.Condition.Min > 0) {
				if(UpdateInterval(reif,reif.MaxVal,reif.MaxVal)) {
					AddChanged(reif);
					return true;
				}
			}
			else if (reif.Condition.Max < 0) {
				if(UpdateInterval(reif,reif.MinVal,reif.MinVal)) {
					AddChanged(reif);
					return true;
				}				
			}
			return false;
		}
		
        public bool Visit(Sum sum) {			
			double min = 0;
			double max = 0;

			for(int i=sum.Terms.Count-1; i>=0; --i) {
				min += sum.Terms[i].Min;
				max += sum.Terms[i].Max;
			}			
			if(UpdateInterval(sum,min,max)) {
				AddChanged(sum);
				return true;
			}
			return false;
		}

        public bool Visit(AutoDiff.Variable variable) {
			return true;
		}

        public bool Visit(Log log) {
			if(UpdateInterval(log,Math.Log(log.Arg.Min),Math.Log(log.Arg.Max))) {
				AddChanged(log);
				return true;
			}
			return false;
		}
        public bool Visit(Sin sin) {
			double size = sin.Arg.Max - sin.Arg.Min;
			bool c = false;
			if (size <= 2*Math.PI) {
				double a = Math.Sin(sin.Arg.Max);
				double b = Math.Sin(sin.Arg.Min);
				double halfPI = Math.PI/2;
				double x = Math.Ceiling((sin.Arg.Min-halfPI) / Math.PI);
				double y = Math.Floor((sin.Arg.Max-halfPI)/ Math.PI);
				if (x==y) { //single extrema
					if(((int)x)%2 == 0) { //maxima
						c = UpdateInterval(sin,Math.Min(a,b),1);
					} else { //minima
						c = UpdateInterval(sin,-1,Math.Max(a,b));
					}						
				} else if (x > y) { //no extrema
					c = UpdateInterval(sin,Math.Min(a,b),Math.Max(a,b));
				} //multiple extrema, don't update
			}
			if(c) AddChanged(sin);
			return c;
		}

        public bool Visit(Cos cos) {
			double size = cos.Arg.Max - cos.Arg.Min;
			bool c = false;
			if (size <= 2*Math.PI) {
				double a = Math.Cos(cos.Arg.Max);
				double b = Math.Cos(cos.Arg.Min);				
				double x = Math.Ceiling(cos.Arg.Min / Math.PI);
				double y = Math.Floor(cos.Arg.Max / Math.PI);
				if (x==y) { //single extrema
					if(((int)x)%2 == 0) { //maxima
						c = UpdateInterval(cos,Math.Min(a,b),1);
					} else { //minima
						c = UpdateInterval(cos,-1,Math.Max(a,b));
					}						
				} else if (x > y) { //no extrema
					c = UpdateInterval(cos,Math.Min(a,b),Math.Max(a,b));
				} //multiple extrema, don't update
			}
			if (c) AddChanged(cos);
			return c;
		}

        public bool Visit(Abs abs) {
			bool containsZero = abs.Arg.Min * abs.Arg.Max <= 0;
			bool c = false;
			if (containsZero) c = UpdateInterval(abs,0, Math.Max(Math.Abs(abs.Arg.Min),Math.Abs(abs.Arg.Max)));
			else c = UpdateInterval(abs,Math.Min(Math.Abs(abs.Arg.Min),Math.Abs(abs.Arg.Max)), Math.Max(Math.Abs(abs.Arg.Min),Math.Abs(abs.Arg.Max)));
			if (c) AddChanged(abs);
			return c;
		}
		
 
        public bool Visit(Exp exp) {
			if(UpdateInterval(exp,Math.Exp(exp.Arg.Min),Math.Exp(exp.Arg.Max))) {
				AddChanged(exp);
				return true;
			}
			return false;
		}
		public bool Visit(Atan2 atan2) {
			throw new NotImplementedException("Atan2 prop not implemented!");
		}
				
		protected void OutputChange(Term t,double oldmin, double oldmax) {
			//Console.WriteLine("UW: Interval of {0} is now [{1}, {2}]",t,t.Min,t.Max);
			double oldwidth = oldmax - oldmin;
			double newwidth = t.Max - t.Min;
			if(t is AutoDiff.Variable) Console.WriteLine("UW shrinking [{0}..{1}] to [{2}..{3}] by {4} ({5}%)",oldmin,oldmax,t.Min,t.Max,oldwidth-newwidth,(oldwidth-newwidth)/oldwidth*100);
		}
		protected bool UpdateInterval(Term t, double min, double max) {
			bool ret = t.Min < min || t.Max > max;
#if DEBUG_UP
			double oldmin = t.Min;
			double oldmax = t.Max;
#endif 		
			if(!Double.IsNaN(min)) t.Min = Math.Max(t.Min,min);
			if(!Double.IsNaN(max)) t.Max = Math.Min(t.Max,max);
			if(ret) IntervalPropagator.updates++;
			IntervalPropagator.visits++;
#if DEBUG_UP
			if (ret) OutputChange(t,oldmin,oldmax);
#endif			
			if (t.Min > t.Max) throw new UnsolveableException();
			return ret;
		}
		
	}
	
}

File: java/src/main/java/Reasoner/IntervalPropagation/UpwardPropagator.java
//
// Translated by CS2J (http://www.cs2j.com): 12.11.2016 09:40:47
//

package Reasoner.IntervalPropagation;


import AutoDiff.ITermVisitor;
import AutoDiff.Product;
import AutoDiff.Term;
import org.apache.commons.lang3.NotImplementedException;
import AutoDiff.*;

public class UpwardPropagator implements ITermVisitor<Boolean>
{
    public TermList Changed;

    private void addChanged(Term t) throws Exception {
        for (Term s : t.Parents)
        {
            if (!Changed.contains(s))
                Changed.enqueue(s);
             
        }
        //Changed.MoveToEnd(s);
        if (!Changed.contains(t))
            Changed.enqueue(t);
         
    }

    public UpwardPropagator() throws Exception {
    }

    public Boolean visit(Constant constant) throws Exception {
        return false;
    }

    public Boolean visit(Zero zero) throws Exception {
        return false;
    }

    public Boolean visit(ConstPower intPower) throws Exception {
        boolean includesZero = intPower.getBase().Min * intPower.getBase().Max <= 0;
        if (intPower.getExponent() > 0)
        {
            double a = Math.pow(intPower.getBase().Max, intPower.getExponent());
            double b = Math.pow(intPower.getBase().Min, intPower.getExponent());
            if (includesZero)
            {
                if (updateInterval(intPower, Math.min(0, Math.min(a, b)), Math.max(0, Math.max(a, b))))
                {
                    addChanged(intPower);
                    return true;
                }
                 
            }
            else
            {
                if (updateInterval(intPower, Math.min(a, b), Math.max(a, b)))
                {
                    addChanged(intPower);
                    return true;
                }
                 
            } 
        }
        else if (!includesZero)
        {
            double a = Math.pow(intPower.getBase().Max, intPower.getExponent());
            double b = Math.pow(intPower.getBase().Min, intPower.getExponent());
            if (updateInterval(intPower, Math.min(a, b), Math.max(a, b)))
            {
                addChanged(intPower);
                return true;
            }
             
        }
          
        return false;
    }

    //else +- Infinity is possible
    public Boolean visit(TermPower tp) throws Exception {
        throw new NotImplementedException("Propagation for TermPower not implemented");
    }


    public Boolean visit(Product product) throws Exception {
        double aa = product.getLeft().Min * product.getRight().Min;
        double bb = product.getLeft().Max * product.getRight().Max;
        double max;
        double min;
        if (product.getLeft() == product.getRight())
        {
            min = Math.min(aa, bb);
            max = Math.max(aa, bb);
            if (product.getLeft().Min * product.getLeft().Max <= 0)
                min = 0;
             
        }
        else
        {
            double ab = product.getLeft().Min * product.getRight().Max;
            double ba = product.getLeft().Max * product.getRight().Min;
            max = Math.max(aa, Math.max(ab, Math.max(ba, bb)));
            min = Math.min(aa, Math.min(ab, Math.min(ba, bb)));
        } 
        if (updateInterval(product, min, max))
        {
            addChanged(product);
            return true;
        }
         
        return false;
    }

    public Boolean visit(LinSigmoid sigmoid) throws Exception {
        throw new NotImplementedException("LinSigmoid for TermPower not implemented");
    }

    public Boolean visit(LTConstraint constraint) throws Exception {
        if (constraint.getLeft().Max < constraint.getRight().Min)
        {
            if (updateInterval(constraint, 1, 1))
            {
                addChanged(constraint);
                return true;
            }
             
        }
        else if (constraint.getLeft().Min >= constraint.getRight().Max)
        {
            //Console.WriteLine("LT UP negated: {0} {1}",constraint.getLeft().Min ,constraint.getRight().Max);
            if (updateInterval(constraint, Double.NEGATIVE_INFINITY, 0))
            {
                addChanged(constraint);
                return true;
            }
             
        }
          
        return false;
    }

    public Boolean visit(LTEConstraint constraint) throws Exception {
        if (constraint.getLeft().Max <= constraint.getRight().Min)
        {
            if (updateInterval(constraint, 1, 1))
            {
                addChanged(constraint);
                return true;
            }
             
        }
        else if (constraint.getLeft().Min > constraint.getRight().Max)
        {
            if (updateInterval(constraint, Double.NEGATIVE_INFINITY, 0))
            {
                addChanged(constraint);
                return true;
            }
             
        }
          
        return false;
    }

    public Boolean visit(Min min) throws Exception {
        if (updateInterval(min, Math.min(min.getLeft().Min, min.getRight().Min), Math.max(min.getLeft().Max, min.getRight().Max)))
        {
            addChanged(min);
            return true;
        }
         
        return false;
    }

    public Boolean visit(Max max) throws Exception {
        if (updateInterval(max, Math.min(max.getLeft().Min, max.getRight().Min), Math.max(max.getLeft().Max, max.getRight().Max)))
        {
            addChanged(max);
            return true;
        }
         
        return false;
    }

    public Boolean visit(And and) throws Exception {
        if (and.getLeft().Min > 0 && and.getRight().Min > 0)
        {
            if (updateInterval(and, 1, 1))
            {
                addChanged(and);
                return true;
            }
             
        }
        else if (and.getLeft().Max <= 0 || and.getRight().Max <= 0)
        {
            if (updateInterval(and, Double.NEGATIVE_INFINITY, 0))
            {
                addChanged(and);
                return true;
            }
             
        }
          
        return false;
    }

    public Boolean visit(Or or) throws Exception {
        if (or.getLeft().Min > 0 || or.getRight().Min > 0)
        {
            if (updateInterval(or, 1, 1))
            {
                addChanged(or);
                return true;
            }
             
        }
        else if (or.getLeft().Max <= 0 && or.getRight().Max <= 0)
        {
            if (updateInterval(or, Double.NEGATIVE_INFINITY, 0))
            {
                addChanged(or);
                return true;
            }
             
        }
          
        return false;
    }

    public Boolean visit(ConstraintUtility cu) throws Exception {
        if (cu.getConstraint().Max < 1)
        {
            if (updateInterval(cu, Double.NEGATIVE_INFINITY, cu.getConstraint().Max))
            {
                addChanged(cu);
                return true;
            }
             
        }
         
        if (updateInterval(cu, Double.NEGATIVE_INFINITY, cu.getUtility().Max))
        {
            addChanged(cu);
            return true;
        }
         
        return false;
    }

    public Boolean visit(Reification reif) throws Exception {
        if (reif.getCondition().Min > 0)
        {
            if (updateInterval(reif, reif.getMaxVal(), reif.getMaxVal()))
            {
                addChanged(reif);
                return true;
            }
             
        }
        else if (reif.getCondition().Max < 0)
        {
            if (updateInterval(reif, reif.getMinVal(), reif.getMinVal()))
            {
                addChanged(reif);
                return true;
            }
             
        }
          
        return false;
    }

    @Override
    public Boolean visit(final Negation r) throws Exception
    {
        return null;
    }

    public Boolean visit(Sum sum) throws Exception {
        double min = 0;
        double max = 0;
        for (int i = sum.getTerms().length - 1;i >= 0;--i)
        {
            min += sum.getTerms()[i].Min;
            max += sum.getTerms()[i].Max;
        }
        if (updateInterval(sum, min, max))
        {
            addChanged(sum);
            return true;
        }
         
        return false;
    }

    public Boolean visit(AutoDiff.Variable variable) throws Exception {
        return true;
    }

    public Boolean visit(Log log) throws Exception {
        if (updateInterval(log, Math.log(log.getArg().Min), Math.log(log.getArg().Max)))
        {
            addChanged(log);
            return true;
        }
         
        return false;
    }

    public Boolean visit(Sin sin) throws Exception {
        double size = sin.getArg().Max - sin.getArg().Min;
        boolean c = false;
        if (size <= 2 * Math.PI)
        {
            double a = Math.sin(sin.getArg().Max);
            double b = Math.sin(sin.getArg().Min);
            double halfPI = Math.PI / 2;
            double x = Math.ceil((sin.getArg().Min - halfPI) / Math.PI);
            double y = Math.floor((sin.getArg().Max - halfPI) / Math.PI);
            if (x == y)
            {
                //single extrema
                if (((int)x) % 2 == 0)
                {
                    //maxima
                    c = updateInterval(sin, Math.min(a, b), 1);
                }
                else
                {
                    //minima
                    c = updateInterval(sin, -1, Math.max(a, b));
                } 
            }
            else if (x > y)
            {
                //no extrema
                c = updateInterval(sin, Math.min(a, b), Math.max(a, b));
            }
              
        }
         
        //multiple extrema, don't update
        if (c)
            addChanged(sin);
         
        return c;
    }

    public Boolean visit(Cos cos) throws Exception {
        double size = cos.getArg().Max - cos.getArg().Min;
        boolean c = false;
        if (size <= 2 * Math.PI)
        {
            double a = Math.cos(cos.getArg().Max);
            double b = Math.cos(cos.getArg().Min);
            double x = Math.ceil(cos.getArg().Min / Math.PI);
            double y = Math.floor(cos.getArg().Max / Math.PI);
            if (x == y)
            {
                //single extrema
                if (((int)x) % 2 == 0)
                {
                    //maxima
                    c = updateInterval(cos, Math.min(a, b), 1);
                }
                else
                {
                    //minima
                    c = updateInterval(cos, -1, Math.max(a, b));
                } 
            }
            else if (x > y)
            {
                //no extrema
                c = updateInterval(cos, Math.min(a, b), Math.max(a, b));
            }
              
        }
         
        //multiple extrema, don't update
        if (c)
            addChanged(cos);
         
        return c;
    }

    public Boolean visit(Abs abs) throws Exception {
        boolean containsZero = abs.getArg().Min * abs.getArg().Max <= 0;
        boolean c = false;
        if (containsZero)
            c = updateInterval(abs, 0, Math.max(Math.abs(abs.getArg().Min), Math.abs(abs.getArg().Max)));
        else
            c = updateInterval(abs, Math.min(Math.abs(abs.getArg().Min), Math.abs(abs.getArg().Max)), Math.max(Math.abs(abs.getArg().Min), Math.abs(abs.getArg().Max))); 
        if (c)
            addChanged(abs);
         
        return c;
    }

    public Boolean visit(Exp exp) throws Exception {
        if (updateInterval(exp, Math.exp(exp.getArg().Min), Math.exp(exp.getArg().Max)))
        {
            addChanged(exp);
            return true;
        }
         
        return false;
    }

    @Override
    public Boolean visit(final Sigmoid sigmoid) throws Exception
    {
        throw new NotImplementedException("Sigmoid Propagation not implemeted");
    }

    public Boolean visit(Atan2 atan2) throws Exception {
        throw new NotImplementedException("atan2 Propagation not implemeted");
    }

    protected void outputChange(Term t, double oldmin, double oldmax) throws Exception {
        //Console.WriteLine("UW: Interval of {0} is now [{1}, {2}]",t,t.Min,t.Max);
        double oldwidth = oldmax - oldmin;
        double newwidth = t.Max - t.Min;
        System.out.println("notImplemented");
//        if (t instanceof AutoDiff.Variable)
//            Console.WriteLine("UW shrinking [{0}..{1}] to [{2}..{3}] by {4} ({5}%)", oldmin, oldmax, t.Min, t.Max, oldwidth - newwidth, (oldwidth - newwidth) / oldwidth * 100);
         
    }

    protected Boolean updateInterval(Term t, double min, double max) throws Exception {
        boolean ret = t.Min < min || t.Max > max;
        if (!Double.isNaN(min))
            t.Min = Math.max(t.Min, min);
         
        if (!Double.isNaN(max))
            t.Max = Math.min(t.Max, max);
         
        if (ret)
            IntervalPropagator.updates++;
         
        IntervalPropagator.visits++;
        if (t.Min > t.Max)
            throw new UnsolveableException();
         
        return ret;
    }

}


